File: app/src/main/java/ir/akhbar/NewsListFragment.java
package ir.akhbar;

import android.os.Bundle;
import android.os.Handler;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class NewsListFragment extends Fragment {

    private ProgressBar progress;
    private ProgressBar searchProgress;
    private RelativeLayout failureView;
    private RecyclerView newsRecycler;
    private EditText searchInput;
    private TextView toolbarTitle;
    private ImageView searchAction;

    private Handler handler;
    private Runnable searchRunnable;

    private Networking networking;

    private String searchQuery = "Iran";

    private final String defaultQuery = "Iran";

    private DatabaseThread databaseThread;

    private NewsDao newsDao;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ApplicationDatabase applicationDatabase = Room.databaseBuilder(getContext(), ApplicationDatabase.class, "application_database")
                .build();
        newsDao = applicationDatabase.getNewsDao();
        databaseThread = new DatabaseThread();
        networking = new Networking();
        handler = new Handler();
        searchRunnable = new Runnable() {
            @Override
            public void run() {
                searchProgress.setVisibility(View.VISIBLE);
                searchAction.setVisibility(View.GONE);
                fetchData(networking, searchQuery);
            }
        };
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.fragment_news_list, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        toolbarTitle = (TextView) view.findViewById(R.id.toolbarTitle);
        searchInput = (EditText) view.findViewById(R.id.searchInput);

        searchProgress = (ProgressBar) view.findViewById(R.id.searchProgress);

        searchAction = (ImageView) view.findViewById(R.id.actionSearch);
        searchAction.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toolbarTitle.setVisibility(View.GONE);
                searchInput.setVisibility(View.VISIBLE);
            }
        });

        searchInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(final Editable s) {
                if (s.toString().isEmpty()) {
                    searchQuery = defaultQuery;
                } else {
                    searchQuery = s.toString();
                }
                handler.removeCallbacks(searchRunnable);
                handler.postDelayed(searchRunnable, 1000);
            }
        });

        progress = (ProgressBar) view.findViewById(R.id.progress);
        failureView = (RelativeLayout) view.findViewById(R.id.failureView);
        newsRecycler = (RecyclerView) view.findViewById(R.id.newsRecycler);
        newsRecycler.setLayoutManager(new LinearLayoutManager(getContext()));
        Button retryButton = (Button) view.findViewById(R.id.retryButton);
        retryButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                failureView.setVisibility(View.GONE);
                progress.setVisibility(View.VISIBLE);
                fetchData(networking, defaultQuery);
            }
        });

        fetchData(networking, defaultQuery);
    }

    private void fetchData(Networking networking, String query) {
        networking.getServer()
                .getNewsList(query, "6fba2629782d465abd2dc5f427223cc0")
                .enqueue(new Callback<ServerResponse>() {
                    @Override
                    public void onResponse(Call<ServerResponse> call, Response<ServerResponse> response) {
                        updateRecyclerView(response.body().getArticles());
                    }

                    @Override
                    public void onFailure(Call<ServerResponse> call, Throwable t) {
                        progress.setVisibility(View.GONE);
                        final List<NewsTable> news = new ArrayList<>();
                        databaseThread.addRunnable(new Runnable() {
                            @Override
                            public void run() {
                                System.out.println("Code 1 " + Thread.currentThread().getName());
                                news.addAll(newsDao.getAllNews());
                                NewsData[] newsDatas = new NewsData[news.size()];
                                if (!news.isEmpty()) {
                                    for (int i = 0; i < news.size(); i++) {
                                        NewsTable newsTable = news.get(i);
                                        NewsData newsData = mapNewsTableToNewsData(newsTable);
                                        newsDatas[i] = newsData;
                                    }
                                    updateRecyclerView(newsDatas);
                                } else {
                                    new Handler(getActivity().getMainLooper()).post(new Runnable() {
                                        @Override
                                        public void run() {
                                            failureView.setVisibility(View.VISIBLE);
                                        }
                                    });
                                }
                            }
                        });
                    }
                });
    }

    private void updateRecyclerView(NewsData[] newsData) {
        progress.setVisibility(View.GONE);
        searchProgress.setVisibility(View.GONE);
        searchAction.setVisibility(View.VISIBLE);
        updateDatabase(newsData);
        NewsAdapter adapter = new NewsAdapter(newsData, new NewsItemClickListener() {
            @Override
            public void onClick(NewsData data) {
                Bundle bundle = new Bundle();
                bundle.putString("newsTitle", data.getNewsTitle());
                bundle.putString("newsDescription", data.getNewsDescription());
                bundle.putString("newsImage", data.getNewsImage());
                bundle.putString("newsUrl", data.getUrl());
                NewsDetailFragment detailFragment = new NewsDetailFragment();
                detailFragment.setArguments(bundle);
                getActivity().getSupportFragmentManager()
                        .beginTransaction()
                        .addToBackStack(null)
                        .replace(R.id.fragmentContainer, detailFragment)
                        .commit();
            }
        });
        newsRecycler.setAdapter(adapter);
    }

    private void updateDatabase(final NewsData[] newsData) {
        databaseThread.addRunnable(new Runnable() {
            @Override
            public void run() {
                List<NewsTable> news = new ArrayList<>();
                for (NewsData newsData : newsData) {
                    NewsTable newsTable = mapNewsDataToNewsTable(newsData);
                    news.add(newsTable);
                }
                List<NewsTable> newsTables = newsDao.getAllNews();
                if (!newsTables.isEmpty()) {
                    newsDao.deleteAllNews(newsTables);
                }
                newsDao.addNewsList(news);
            }
        });
    }

    private NewsTable mapNewsDataToNewsTable(NewsData newsData) {
        return new NewsTable(newsData.getNewsTitle(), newsData.getNewsDescription(), newsData.getNewsImage(), newsData.getUrl());
    }

    private NewsData mapNewsTableToNewsData(NewsTable newsTable) {
        return new NewsData(newsTable.getTitle(), newsTable.getDescription(), newsTable.getUrlToImage(), newsTable.getUrl());
    }

    public boolean canHandleBackPressed() {
        boolean canHandleBackPressed = false;
        if (searchInput.getVisibility() == View.VISIBLE) {
            searchInput.setVisibility(View.GONE);
            toolbarTitle.setVisibility(View.VISIBLE);
            searchProgress.setVisibility(View.VISIBLE);
            searchAction.setVisibility(View.GONE);
            fetchData(networking, defaultQuery);
            canHandleBackPressed = true;
        }
        return canHandleBackPressed;
    }

    @Override
    public void onDestroyView() {
        if (searchInput.getVisibility() == View.VISIBLE) {
            searchInput.setVisibility(View.GONE);
            toolbarTitle.setVisibility(View.VISIBLE);
        }
        super.onDestroyView();
    }
}

File: app/src/main/java/ir/akhbar/NewsListFragment.kt
package ir.akhbar

import android.os.Bundle
import android.os.Handler
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.room.Room

import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import kotlinx.android.synthetic.main.fragment_news_list.*

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class NewsListFragment : Fragment() {

    private lateinit var handler: Handler
    private lateinit var searchRunnable: Runnable

    private lateinit var networking: Networking

    private var searchQuery = "Iran"

    private val defaultQuery = "Iran"

    private lateinit var databaseThread: DatabaseThread

    private lateinit var newsDao: NewsDao

    private lateinit var updateDatabaseTask: UpdateDatabaseTask

    private lateinit var queryDatabaseTask: QueryDatabaseTask

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val applicationDatabase = Room.databaseBuilder(
                requireContext(),
                ApplicationDatabase::class.java,
                "application_database"
        ).build()
        newsDao = applicationDatabase.newsDao
        databaseThread = DatabaseThread()
        networking = Networking()
        handler = Handler()
        searchRunnable = Runnable {
            searchProgress.visibility = View.VISIBLE
            actionSearch.visibility = View.GONE
            fetchData(networking, searchQuery)
        }

        updateDatabaseTask = UpdateDatabaseTask(newsDao)
        queryDatabaseTask = QueryDatabaseTask(object : QueryDatabaseCallback {
            override fun onQuerySuccess(newsDatas: Array<NewsData>) {
                updateRecyclerView(newsDatas)
            }

            override fun onQueryFailure() {
                failureView.visibility = View.VISIBLE
            }
        }, newsDao)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.fragment_news_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        actionSearch.setOnClickListener {
            toolbarTitle.visibility = View.GONE
            searchInput.visibility = View.VISIBLE
        }

        searchInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
            }
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            }
            override fun afterTextChanged(s: Editable) {
                searchQuery = if (s.toString().isEmpty()) {
                    defaultQuery
                } else {
                    s.toString()
                }
                handler.removeCallbacks(searchRunnable)
                handler.postDelayed(searchRunnable, 1000)
            }
        })

        newsRecycler.layoutManager = LinearLayoutManager(context)
        retryButton.setOnClickListener {
            failureView.visibility = View.GONE
            progress.visibility = View.VISIBLE
            fetchData(networking, defaultQuery)
        }

        fetchData(networking, defaultQuery)
    }

    private fun fetchData(networking: Networking, query: String) {
        networking.server
                .getNewsList(query, "6fba2629782d465abd2dc5f427223cc0")
                .enqueue(object : Callback<ServerResponse> {
                    override fun onResponse(call: Call<ServerResponse>, response: Response<ServerResponse>) {
                        updateRecyclerView(response.body()!!.articles)
                    }

                    override fun onFailure(call: Call<ServerResponse>, t: Throwable) {
                        progress!!.visibility = View.GONE
                        queryDatabaseTask.execute()
                    }
                })
    }

    private fun updateRecyclerView(newsData: Array<NewsData>) {
        progress.visibility = View.GONE
        searchProgress.visibility = View.GONE
        actionSearch.visibility = View.VISIBLE
        updateDatabase(newsData)
        val adapter = NewsAdapter(newsData, NewsItemClickListenerKt { data ->
            val bundle = Bundle()
            bundle.putString("newsTitle", data.newsTitle)
            bundle.putString("newsDescription", data.newsDescription)
            bundle.putString("newsImage", data.newsImage)
            bundle.putString("newsUrl", data.url)
            val detailFragment = NewsDetailFragment()
            detailFragment.arguments = bundle
            requireActivity().supportFragmentManager
                    .beginTransaction()
                    .addToBackStack(null)
                    .replace(R.id.fragmentContainer, detailFragment)
                    .commit()
        })
        newsRecycler.adapter = adapter
    }

    private fun updateDatabase(newsData: Array<NewsData>) {
        updateDatabaseTask.execute(*newsData)
    }

    private fun mapNewsTableToNewsData(newsTable: NewsTable): NewsData {
        return NewsData(newsTable.title, newsTable.description, newsTable.urlToImage, newsTable.url)
    }

    fun canHandleBackPressed(): Boolean {
        var canHandleBackPressed = false
        if (searchInput.visibility == View.VISIBLE) {
            searchInput.visibility = View.GONE
            toolbarTitle.visibility = View.VISIBLE
            searchProgress.visibility = View.VISIBLE
            actionSearch.visibility = View.GONE
            fetchData(networking, defaultQuery)
            canHandleBackPressed = true
        }
        return canHandleBackPressed
    }

    override fun onDestroyView() {
        if (searchInput.visibility == View.VISIBLE) {
            searchInput.visibility = View.GONE
            toolbarTitle.visibility = View.VISIBLE
        }
        super.onDestroyView()
    }
}
